tsl/src/remote/dist_commands.c

```diff
@@ -76,6 +86,287 @@ ts_dist_cmd_collect_responses(List *requests)
+			break;
+		case T_List:
+			/* Already in the format we want */
+			data_node_name_list_check_acl(data_nodes, ACL_NO_CHECK);
+			break;
+		default:
+			elog(ERROR, "invalid list type %u", nodeTag(data_nodes));
+			break;
+	}
+	return data_nodes;
+}
+
+/*
+ * Callback to manage memory allocated by libpq via malloc
+ *
+ * If a PGresult does not belong to a connection or a connection isn't stored
+ * in the connection cache, all memory associated needs to be handled when the
+ * Memory Context is deleted/reset.
+ */
+static void
+ts_dist_cmd_mctx_reset_callback(void *arg)
+{
+	Assert(arg != NULL);
+	ListCell *lc;
+	List *callback_list = *(List **) arg;
+
+	foreach (lc, callback_list)
+	{
+		DistCmdResetCallback *cb = lfirst(lc);
+		if (cb->result != NULL)
+			PQclear(cb->result);
+		if (cb->conn != NULL)
+			remote_connection_close(cb->conn);
+		pfree(cb);
+	}
+	list_free(callback_list);
+
+	pfree(arg);
+}
+
+/*
+ * Helper function that errors in case of OOM.
+ *
+ */
+static PGresult *
+ts_dist_cmd_make_pg_result(ExecStatusType status)
+{
+	PGresult *pg_result = PQmakeEmptyPGresult(NULL, status);
+	if (pg_result == NULL)
+	{
+		/* If PQmakeEmptyPGresult returns NULL, we are OOM
+		 * There's nothing left to do but to exit now */
+		ereport(ERROR,
+				(errcode(ERRCODE_OUT_OF_MEMORY),
+				 errmsg("out of memory"),
+				 errdetail("Failed to allocate pg_result")));
+	}
+
+	Assert(pg_result != NULL);
+	return pg_result;
+}
+
+static void
+ts_dist_cmd_create_error_response(DistCmdResponse *rsp, const char *err_msg, const char *data_node,
+								  DistCmdResetCallback *cb)
+{
+	Assert(rsp != NULL);
+
+	if (rsp->pg_result == NULL)
+	{
+		Assert(cb != NULL);
+		PGresult *res = ts_dist_cmd_make_pg_result(PGRES_FATAL_ERROR);
+		cb->result = res;
+		rsp->pg_result = res;
+	}
+
+	/* Todo: There is no function to set errorMessage in a pg_result.
+	 * For now, copy errorMessage around in DistCmdResponse */
+	rsp->errorMessage = pstrdup(err_msg);
+	rsp->data_node = pstrdup(data_node);
+}
+
+/*
+ * Invoke SQL statement on the given data nodes without throwing.
+ *
+ * For each broken connection, a PGresult with PGRES_FATAL_ERROR is created.
+ */
+DistCmdResult *
+ts_dist_cmd_invoke_on_data_nodes_no_throw(const char *sql, const char *search_path,
+										  List *data_nodes)
+{
+	ListCell *lc;
+	List *requests = NIL;
+	List **callback_data;
+	DistCmdResult *results;
+	AsyncRequestSet *rs;
+	AsyncResponse *ar;
+	MemoryContextCallback *mcb;
+	bool set_search_path = search_path != NULL;
+
+	if (data_nodes == NIL)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("no data nodes to execute command on"),
+				 errhint("Add data nodes before executing a distributed command.")));
+
+	data_nodes = ts_dist_cmd_sanitize_data_node_list(data_nodes);
+
+	/* Since libpq uses malloc, register a callback that frees results
+	 * on Memory Context Reset/Delete */
+	callback_data = palloc0(sizeof(List *));
```

Comment:
That's actually not the problem here. It should be `sizeof(List **)`. A list is allowed to change its initial pointer when using `lappend()`. Since this is a list which needs to end up in a callback, I need to store the memory address to the list somewhere.
Still, good catch. The intent is more readable now.